extensions/amp-viewer-integration/0.1/messaging.js

```diff
@@ -0,0 +1,193 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+import {listen} from '../../../src/event-helper';
+import {dev} from '../../../src/log';
+
+const sentinel_ = '__AMPHTML__';
+const requestSentinel_ = sentinel_ + 'REQUEST';
+const responseSentinel_ = sentinel_ + 'RESPONSE';
+
+/**
+ * @fileoverview This is used in amp-viewer-integration.js for the
+ * communication protocol between AMP and the viewer. In the comments, I will
+ * refer to the communication as a conversation between me and Bob. The
+ * messaging protocol should support both sides, but at this point I'm the
+ * ampdoc and Bob is the viewer.
+ */
+export class Messaging {
+
+  /**
+   * Conversation (messaging protocol) between me and Bob.
+   * @param {!Window} source
+   * @param {!Window} target
+   * @param {string} targetOrigin
+   * @param {function(string, *, boolean):(!Promise<*>|undefined)}
+   *    requestProcessor
+   */
+  constructor(source, target, targetOrigin, requestProcessor) {
+    /**  @private {!number} */
+    this.requestIdCounter_ = 0;
+    /**  @private {*} */
+    this.waitingForResponse_ = {};
+    /** @const @private {!Window} */
+    this.target_ = target;
+    /** @const @private {string} */
+    this.targetOrigin_ = targetOrigin;
+    /** @const @private {function(string, *, boolean):(!Promise<*>|undefined)} */
+    this.requestProcessor_ = requestProcessor;
+
+    dev().assert(this.targetOrigin_, 'Target origin must be specified!');
+
+    listen(source, 'message', this.handleMessage_.bind(this));
+  }
+
+
+  /**
+   * Bob sent me a message. I need to decide if it's a new request or
+   * a response to a previous 'conversation' we were having.
+   * @param {Event|null} event
+   * @private
+   */
+  handleMessage_(event) {
+    if (!event || event.source != this.target_ ||
+      event.origin != this.targetOrigin_) {
+      return;
+    }
+    const message = event.data;
+    if (message.sentinel == requestSentinel_) {
+      this.handleRequest_(message);
+    } else if (message.sentinel == responseSentinel_) {
+      this.handleResponse_(message);
+    }
+  }
+
+
+  /**
+   * I'm sending Bob a new outgoing request.
+   * @param {string} eventType
+   * @param {*} payload
+   * @param {boolean} awaitResponse
+   * @return {!Promise<*>|undefined}
+   */
+  sendRequest(eventType, payload, awaitResponse) {
+    dev().info(
+        'MESSAGING', 'messaging.js -> sendRequest, eventType: ', eventType);
+    const requestId = ++this.requestIdCounter_;
+    let promise = undefined;
+    if (awaitResponse) {
+      promise = new Promise((resolve, reject) => {
+        this.waitingForResponse_[requestId] = {resolve, reject};
+      });
+    }
+    this.sendMessage_(requestSentinel_, requestId.toString(), eventType,
+      payload, awaitResponse);
+    return promise;
+  }
+
+
+  /**
+   * I'm responding to a request that Bob made earlier.
+   * @param {number} requestId
```

Comment:
We need to decide whether `requestId` is a `number` or a `string`. It's used as both in different places. I think it should be a `string`.